datafusion/physical-expr/src/aggregate/count.rs

```diff
@@ -76,6 +85,109 @@ impl Count {
     }
 }
 
+/// An accumulator to compute the counts of [`PrimitiveArray<T>`].
+/// Stores values as native types, and does overflow checking
+///
+/// Unlike most other accumulators, COUNT never produces NULLs. If no
+/// non-null values are seen in any group the output is 0. Thus, this
+/// accumulator has no additional null or seen filter tracking.
+#[derive(Debug)]
+struct CountGroupsAccumulator {
+    /// Count per group (use i64 to make Int64Array)
+    counts: Vec<i64>,
+}
+
+impl CountGroupsAccumulator {
+    pub fn new() -> Self {
+        Self { counts: vec![] }
+    }
+}
+
+impl GroupsAccumulator for CountGroupsAccumulator {
+    fn update_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indices: &[usize],
+        opt_filter: Option<&arrow_array::BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 1, "single argument to update_batch");
+        let values = values.get(0).unwrap();
+
+        // Add one to each group's counter for each non null, non
+        // filtered value
+        self.counts.resize(total_num_groups, 0);
+        accumulate_indices(
+            group_indices,
+            values.nulls(), // ignore values
+            opt_filter,
+            |group_index| {
```

Comment:
Should this use wrapping add as well?

Reply:
I don't know what the implications are -- would that be faster (but not check overflow)? I am happy to change it but I honestly don't know

Reply:
I suggested it mainly for consistency.
AFAIK `wrapping_add` doesn't check for overflow but wraps, which is the default already, but in debug mode it causes an exception.
Maybe we should remove the `wrapping_add` in the other places and "just" use `+` etc.

Reply:
When I tried to use `wrapping_add` then `rustc` told me I needed to check the return type. Given I never expect an overflow and this is on the hot path, I think we should not use `wrapping_add`

```
warning: unused return value of `core::num::<impl i64>::wrapping_add` that must be used
   --> datafusion/physical-expr/src/aggregate/count.rs:130:17
    |
130 |                 self.counts[group_index].wrapping_add(1);
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this returns the result of the operation, without modifying the original
    = note: `#[warn(unused_must_use)]` on by default
help: use `let _ = ...` to ignore the resulting value
    |
130 |                 let _ = self.counts[group_index].wrapping_add(1);
    |                 +++++++
```

> Maybe we should remove the wrapping_add in the other places and "just" use + etc.

What other places are you referring to? I didn't see any in this PR but maybe I am misunderstanding

Reply:
We're using `add_wrapping` in sum, average:

`sum.add_wrapping(new_value)` etc.

Reply:
Ah, I was confused -- I used `add_wrapping` (part of `std::ops`) rather than `wrapping_add` (part of Arrow NativeType) 🤦 

I played around with it -- it turns out anything with an actual rust type (like i64) I can use `+=` but for templated code on `ArrowNativeType` I can't use `+=` I have to use `wrapping_add`

This is how the code would look like if I use `add_wrapping`, which I think looks a bit more tortured (though it is consistent). Do you think I should make the change?

```diff
diff --git a/datafusion/physical-expr/src/aggregate/count.rs b/datafusion/physical-expr/src/aggregate/count.rs
index 287970de29..cfb1713d8e 100644
--- a/datafusion/physical-expr/src/aggregate/count.rs
+++ b/datafusion/physical-expr/src/aggregate/count.rs
@@ -156,9 +156,10 @@ impl GroupsAccumulator for CountGroupsAccumulator {
                 .iter()
                 .zip(group_indices.iter())
                 .zip(partial_counts.iter())
-                .for_each(|((filter_value, &group_index), partial_count)| {
+                .for_each(|((filter_value, &group_index), &partial_count)| {
                     if let Some(true) = filter_value {
-                        self.counts[group_index] += partial_count;
+                        let count = &mut self.counts[group_index];
+                        *count = count.add_wrapping(partial_count);
                     }
                 }),
             None => group_indices.iter().zip(partial_counts.iter()).for_each(
```